### PR TITLE
Fix build

### PR DESCRIPTION
**Pull request by mvz**
This set of commits makes acts_as_paranoid's build pass on activerecord 4.0 and 4.1.

There is one incompatible change, which is that `#destroy!` now does the same as `#destroy`. This is because AR assumes that the only potential difference is that `#destroy!` might raise an exception and calls that version in some callbacks.

### CODE REVIEW

**Comment by jodosha** [comment]
Please merge this :+1: It also solves this problem: https://github.com/ActsAsParanoid/acts_as_paranoid/issues/29

File: lib/acts_as_paranoid/core.rb [hunk 1]
```diff
@@ -112,10 +118,14 @@ def destroy
           end
         end
       else
-        destroy!
+        destroy_fully!
       end
     end
 
+    def destroy
+      destroy!
```
Comment (zzak) [inline lib/acts_as_paranoid/core.rb]: This change I'm particularly weary of, it's going to be a big compatibility change in many cases.. and could break some applications.

Comment (jodosha) [inline lib/acts_as_paranoid/core.rb]: @zzak I'm sorry, I haven't noticed that method.
@mvz Is there any reason why this is implemented like that?

Is there anything that I can help with? Thank you all.

Comment (jodosha) [inline lib/acts_as_paranoid/core.rb]: @zzak @mvz Sorry for the stupid question, as it's been long time since I dug inside AR internals.

Correct me if I'm wrong. That `#destroy` method uses internally `#destroy!`, which isn't the one from `ActiveRecord::Base`, but the one implemented a few lines above in the source file.

That `AAP#destroy!` updates the `deleted_at` column with the current timestamp. It doesn't remove the record from the database (like `AR::Base` does).

This last behavior is implemented by `AAP#destroy_fully!`.

To recap: by calling `AAP#destroy` we get a soft delete, which seems to be legit.

@zzak What do you think?

Comment (mvz) [inline lib/acts_as_paranoid/core.rb]: @jodosha the reason is that ActiveRecord currently assumes that `#destroy` and `#destroy!` do basically the same thing, except the latter raises exceptions. Because of this, it thinks it's safe to call `#destroy!` on dependent objects when the user calls `#destroy` on the parent object. However, AAP uses the `!` to signify a hard delete. I resolved this conflict by implementing the `#destroy_fully` method. See also the commit comment here: 3c12279a093eeebe22596a9db92eacb19ca5e62d

Comment (zzak) [inline lib/acts_as_paranoid/core.rb]: It makes sense after reading the commit comment mentioned above.

Comment (jodosha) [inline lib/acts_as_paranoid/core.rb]: @zzak If there isn't any blocker, would you like to merge this? :yellow_heart: :green_heart: :heart: :blue_heart: :heart: Thank you!

**Comment by defeated** [comment]
This looks great. Could we also get a minor edit to __Real deletion__ section to reflect the API change?

https://github.com/ActsAsParanoid/acts_as_paranoid#real-deletion

replacing `destroy!` with `destroy_fully!`

**Comment by metacortex** [comment]
:+1:

**Comment by whatasunnyday** [comment]
@zzak Hey sorry to ping you know but any chance we can see this merged?

**Comment by mvz** [comment]
Thanks, @zzak!

**Comment by zzak** [comment]
@mvz No, THANK YOU!! <3

**Comment by mvz** [comment]
LOL!

**Comment by jodosha** [comment]
@mvz @zzak Thank YOU :green_heart:
